File: global_business/business/transaction.py
from datetime import datetime


class Transaction:
    def __init__(self, account: "Account", transaction_type: str, amount: float):
        self.account = account
        self.transaction_type = transaction_type
        self.amount = amount
        self.timestamp = datetime.now()

    def process(self):
        if self.transaction_type == "deposit":
            self.account.deposit(self.amount)
        elif self.transaction_type == "withdrawal":
            self.account.withdraw(self.amount)
        else:
            raise ValueError("Invalid transaction type")

File: global_business/business/models/transaction.rb
class Transaction < ApplicationRecord
  # Encryption and decryption of sensitive data
  encrypts :amount

  # Associations
  belongs_to :account
  belongs_to :user

  # Validations
  validates :amount, presence: true, numericality: { greater_than: 0.01, less_than: 100000.00 }
  validates :transaction_type, presence: true, inclusion: { in: %w[deposit withdrawal] }

  # Callbacks
  before_save :hash_amount
  after_create :send_transaction_email

  private

  def hash_amount
    self.amount = Encryption.hash(self.amount)
  end

  def send_transaction_email
    TransactionMailer.transaction_email(self).deliver_now
  end
end
